pallets/xtransfer/src/bridge_transfer/mod.rs

```diff
@@ -0,0 +1,278 @@
+	pub enum Event<T: Config> {
+		/// [chainId, min_fee, fee_scale]
+		FeeUpdated(bridge::BridgeChainId, BalanceOf<T>, u32),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		AssetConversionFailed,
+		AssetNotRegistered,
+		FeeOptionsMissing,
+		InvalidDestination,
+		InvalidFeeOption,
+		InsufficientBalance,
+		BalanceConversionFailed,
+		FailedToTransactAsset,
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn bridge_fee)]
+	pub type BridgeFee<T: Config> =
+		StorageMap<_, Twox64Concat, bridge::BridgeChainId, (BalanceOf<T>, u32), ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Change extra bridge transfer fee that user should pay
+		#[pallet::weight(195_000_000)]
+		pub fn change_fee(
+			origin: OriginFor<T>,
+			min_fee: BalanceOf<T>,
+			fee_scale: u32,
+			dest_id: bridge::BridgeChainId,
+		) -> DispatchResult {
+			T::BridgeCommitteeOrigin::ensure_origin(origin)?;
+			ensure!(fee_scale <= 1000u32, Error::<T>::InvalidFeeOption);
+			BridgeFee::<T>::insert(dest_id, (min_fee, fee_scale));
+			Self::deposit_event(Event::FeeUpdated(dest_id, min_fee, fee_scale));
+			Ok(())
+		}
+
+		/// Transfer some amount of specific asset to some recipient on a (whitelisted) distination chain.
+		#[pallet::weight(195_000_000)]
+		#[transactional]
+		pub fn transfer_assets(
+			origin: OriginFor<T>,
+			asset: XTransferAsset,
+			dest_id: bridge::BridgeChainId,
+			recipient: Vec<u8>,
+			amount: BalanceOf<T>,
+		) -> DispatchResult {
+			let source = ensure_signed(origin)?;
+			ensure!(
+				<bridge::Pallet<T>>::chain_whitelisted(dest_id),
+				Error::<T>::InvalidDestination
+			);
+			ensure!(
+				BridgeFee::<T>::contains_key(&dest_id),
+				Error::<T>::FeeOptionsMissing
+			);
+			let asset_id: <T as pallet_assets::Config>::AssetId =
+				T::AssetsWrapper::id(&asset).ok_or(Error::<T>::AssetNotRegistered)?;
+			let asset_amount = T::BalanceConverter::to_asset_balance(amount, asset_id)
+				.map_err(|_| Error::<T>::BalanceConversionFailed)?;
+			ensure!(
+				<pallet_assets::pallet::Pallet<T>>::balance(asset_id.into(), &source)
+					>= asset_amount,
+				Error::<T>::InsufficientBalance
+			);
+
+			let rid: bridge::ResourceId = asset
+				.try_into()
+				.map_err(|_| Error::<T>::AssetConversionFailed)?;
+			let fee = Self::calculate_fee(dest_id, amount);
+			// check native balance to cover fee
+			let native_free_balance = <T as Config>::Currency::free_balance(&source);
+			ensure!(native_free_balance >= fee, Error::<T>::InsufficientBalance);
+
+			// pay fee to treasury
+			let imbalance = <T as Config>::Currency::withdraw(
+				&source,
+				fee,
+				WithdrawReasons::FEE,
+				ExistenceRequirement::AllowDeath,
+			)?;
+			T::OnFeePay::on_unbalanced(imbalance);
+
+			let asset_amount = T::BalanceConverter::to_asset_balance(amount, asset_id)
+				.map_err(|_| Error::<T>::BalanceConversionFailed)?;
+			// burn asset from sender
+			<pallet_assets::pallet::Pallet<T> as FungibleMutate<T::AccountId>>::burn_from(
```

Comment:
Right, with x3(evm -> khala -> parachains), things become more complicated, i also think about how it can fit the x3 security situtions. Maybe give a solution in the upcoming x3 PR.

Reply:
All bridge transfer are now reserve based, see [here](https://github.com/Phala-Network/khala-parachain/pull/47/commits/04de2e699d194a6aaf7c2e69b7bd2e1aee14cd2e)

Reply:
Cool. We can merge this and review the x3 PR later